packages/cojson/src/sync.ts

```diff
@@ -708,35 +708,61 @@ export class SyncManager {
     }
 
     for (const peer of this.getPeers()) {
-      this.syncStateSubscriptionManager.triggerUpdate(peer.id, coValue.id);
+      this.syncState.triggerUpdate(peer.id, coValue.id);
     }
   }
 
-  async waitForUploadIntoPeer(peerId: PeerID, id: RawCoID) {
-    const isAlreadyUploaded =
-      this.syncStateSubscriptionManager.getIsCoValueFullyUploadedIntoPeer(
-        peerId,
-        id,
-      );
+  async waitForSyncWithPeer(peerId: PeerID, id: RawCoID, timeout = 10000) {
+    const { syncState } = this;
+    const currentSyncState = syncState.getCurrentSyncState(peerId, id);
 
-    if (isAlreadyUploaded) {
+    const isTheConditionAlreadyMet = currentSyncState.uploaded;
+
+    if (isTheConditionAlreadyMet) {
       return true;
     }
 
-    return new Promise((resolve) => {
-      const unsubscribe =
-        this.syncStateSubscriptionManager.subscribeToPeerUpdates(
-          peerId,
-          (knownState, syncState) => {
-            if (syncState.isUploaded && knownState.id === id) {
-              resolve(true);
-              unsubscribe?.();
-            }
-          },
-        );
+    return new Promise((resolve, reject) => {
+      const unsubscribe = this.syncState.subscribeToPeerUpdates(
+        peerId,
+        (knownState, syncState) => {
+          if (syncState.uploaded && knownState.id === id) {
+            resolve(true);
+            unsubscribe?.();
+            clearTimeout(timeoutId);
+          }
+        },
+      );
+
+      const timeoutId = setTimeout(() => {
+        reject(new Error(`Timeout waiting for sync on ${peerId}/${id}`));
+        unsubscribe?.();
+      }, timeout);
     });
   }
 
+  async waitForSync(id: RawCoID, timeout = 30_000) {
+    const peers = this.getPeers();
+
+    return Promise.all(
+      peers
+        .filter((peer) => peer.role !== "storage")
```

Comment:
Removed the storage peers because we are not getting an ACK for the sent content.

Reply:
Fixed by changing PeerState to share the same instance of "optimisticKnownState" and "knownState" when the role is Peer.